src/dijkstra.rs

```diff
@@ -0,0 +1,159 @@
+use std::collections::{ BinaryHeap, HashMap };
+use std::cmp::Ordering;
+
+use weighted_graph::{ Graph, NodeId };
+
+pub fn shortest_path(graph: &Graph,
+                     source: &str,
+                     destination: Option<&str>
+                    ) -> (i64, HashMap<NodeId, Result>) {
+
+    let mut min_heap = BinaryHeap::new();
+    let mut results = HashMap::new();
+
+    let initial = Result { id: source.to_string(), cost: 0, predecessor: "".to_string() };
+    results.insert(source.to_string(), initial.clone());
+    min_heap.push(initial.clone());
+
+    while let Some(current) = min_heap.pop() {
+        if let Some(target) = destination {
+            if current.id == target {
+                return (current.cost, results)
+            }
+        }
+
+        if let Some(edges) = graph.get_edges(&current.id) {
+            let mut iter = edges.iter();
+            while let Some(edge) = iter.next() {
+                match graph.get_node(&edge.to_id) {
+                    Some(node) => {
+                        let node_cost = results.get(&node.id)
+                                               .map_or(i64::max_value(), |node| node.cost);
+                        if current.cost + edge.weight < node_cost {
+                            let hnode = Result { id: node.id.clone(),
+                                                cost: current.cost + edge.weight,
+                                                predecessor: current.id.clone()
+                                              };
+                            min_heap.push(hnode.clone());
+                            results.insert(node.id.clone(), hnode.clone());
+                        }
+                    }
+                    None => {}
+                }
+            }
+        }
+    }
+    (0, results)
+}
+
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct Result {
```

Comment:
FWIW `Result<T>` is usually a typedef for `std::result::Result<T, package_errtype>`.